[Unity] nn.Module external modules

cyx-6 opened the pull request:
This PR introduces the feature of importing external `*.o` modules into our `nn.Module` frontend.

junrushao commented:
I'd love to review later this week. Was meant to review Monday but got distracted for quite a while

cyx-6 commented on python/tvm/relax/frontend/nn/torch.py:
```diff
@@ -42,6 +42,7 @@ def __init__(  # pylint: disable=invalid-name
         spec: _spec.ModuleSpec,
         vm: VirtualMachine,
         params: List[NDArray],
+        effects: List[Any] = None,
```
yes

junrushao commented on python/tvm/relax/frontend/nn/spec.py:
```diff
@@ -268,10 +306,24 @@ def _effects() -> List[Tuple[str, core.Effect]]:
+                spec.module, "", condition_yield=lambda x: isinstance(x, core.ExternModule)
+            ):
+                module_spec = extern_module.module_spec
+                if not module_spec.filename in mod2func:
+                    mod2func[module_spec.filename] = set()
```
using

```
from collections import defaultdict

mod2func = defaultdict(set)
```

to avoid checking if `mod2func` contains `module_spec.filename`.

junrushao commented on python/tvm/relax/frontend/nn/spec.py:
```diff
@@ -268,10 +306,24 @@ def _effects() -> List[Tuple[str, core.Effect]]:
+                mod2func[module_spec.filename].update(
+                    [function_spec.symbol for function_spec in module_spec.functions]
+                )
+            return [(mod, list(mod2func[mod])) for mod in mod2func]
```
always prefer `[f(k, v) for k, v in dict.items()]` to `[f(k, dict[k]) for k in dict]`

junrushao commented on python/tvm/relax/frontend/nn/spec.py:
```diff
@@ -282,7 +334,16 @@ def _effects() -> List[Tuple[str, core.Effect]]:
+        mod = self.builder.get()
+        if extern_modules:
+            original_external_mods = mod.get_attr("external_mods")
+            if original_external_mods is not None:
+                external_mods.extend(original_external_mods)
```
let's follow the ordering and put the original ones in the beginning.

```
external_mods = list(original_external_mods) + external_mods
```

junrushao commented on tests/python/relax/test_frontend_nn_extern_module.py:
```diff
@@ -0,0 +1,122 @@
+    ],
+)
+def test_extern_module(shape_a, shape_b, shape_out, dtype, value_table, var_table):
+    def shape_converter(shape, table):
+        out_shape = []
+        for value in shape:
+            if isinstance(value, str):
+                out_shape.append(table[value])
+            else:
+                out_shape.append(value)
+        return tuple(out_shape)
+
+    Module = get_relax_matmul_module(
+        shape_converter(shape_a, var_table),
+        shape_converter(shape_b, var_table),
+        dtype,
+    )
+
+    mod = partition_for_cutlass(Module, True)
+    codegen_pass = relax.transform.RunCodegen({"cutlass": {"sm": 80, "find_first_valid": True}})
+    mod = codegen_pass(mod)
```
iiuc this part is only used for generating `./tmp/cutlass.o` - is that correct?

if so, is it possible to:
1) use a much simpler and more unit-testable source code? for example:

```C++
void Add(NDArray a, NDArray b, NDArray c) {
  ICHECK(a->ndim == 0 && a->dtype == DataType::Float(32));
  ICHECK(b->ndim == 0 && b->dtype == DataType::Float(32));
  ICHECK(c->ndim == 0 && c->dtype == DataType::Float(32));
  *(reinterpret_cast<double*>(c->data)) = *(reinterpret_cast<double*>(a->data)) + *(reinterpret_cast<double*>(b->data));
}
```

2) isolate this logic to a separate method, say, "def generate_test_module(path: str)"

junrushao commented:
Otherwise LGTM

cyx-6 committed:
[Unity] nn.Module external modules

This PR introduces the feature of importing external `*.o` modules into our `nn.Module` frontend.

Commit:
apply code review suggestions

Commit:
add ASF header

Cydia2018 commented:
Thank you very much for your work. 

I would like to ask if the function of external modules has not been fully developed yet? I tried to embed a handwritten attention operator and it compiled normally, but an error was reported when running: Cannot find PackedFunc attention in either Relax VM kernel library.

Looking forward to your reply!

cyx-6 commented:
@Cydia2018 Thanks for reaching out! Have you ever wrapped your operator with `TVM_DLL_EXPORT_TYPED_FUNC` or other TVM registration functor? Feel free to create a new issue with more context. And it would be great if you could provide a reproducible script for us. :)
